pkg/loop/plugin_median.go

```diff
@@ -37,116 +39,111 @@ func PluginMedianHandshakeConfig() plugin.HandshakeConfig {
-	return &plugin.ClientConfig{
-		HandshakeConfig: PluginMedianHandshakeConfig(),
-		Plugins: map[string]plugin.Plugin{
-			PluginMedianName: NewGRPCPluginMedian(nil, lggr),
-		},
-		AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
-	}
-}
-
-type grpcPluginMedian struct {
+type GRPCPluginMedian struct {
 	plugin.NetRPCUnsupportedPlugin
 
-	impl PluginMedian
-	lggr logger.Logger
-}
+	StopCh <-chan struct{}
+	Logger logger.Logger
 
-func NewGRPCPluginMedian(mp PluginMedian, lggr logger.Logger) plugin.Plugin {
-	return &grpcPluginMedian{impl: mp, lggr: lggr}
+	PluginServer PluginMedian
+
+	pluginClient *pluginMedianClient
 }
 
-func (p *grpcPluginMedian) GRPCServer(broker *plugin.GRPCBroker, server *grpc.Server) error {
-	pb.RegisterPluginMedianServer(server, newPluginMedianServer(&lggrBroker{p.lggr, broker}, p.impl))
-	return nil
+func (p *GRPCPluginMedian) GRPCServer(broker *plugin.GRPCBroker, server *grpc.Server) error {
+	pb.RegisterServiceServer(server, &serviceServer{srv: p.PluginServer})
+	pb.RegisterPluginMedianServer(server, newPluginMedianServer(&lggrBroker{p.StopCh, p.Logger, broker}, p.PluginServer))
+	ctx, cancel := utils.ContextFromChan(p.StopCh)
+	defer cancel()
+	return p.PluginServer.Start(ctx)
 }
 
 // GRPCClient implements [plugin.GRPCPlugin] and returns a [PluginRelayer].
-func (p *grpcPluginMedian) GRPCClient(ctx context.Context, broker *plugin.GRPCBroker, conn *grpc.ClientConn) (interface{}, error) {
-	return newMedianPluginClient(p.lggr, broker, conn), nil
+func (p *GRPCPluginMedian) GRPCClient(_ context.Context, broker *plugin.GRPCBroker, conn *grpc.ClientConn) (interface{}, error) {
+	if p.pluginClient == nil {
+		p.pluginClient = newPluginMedianClient(p.StopCh, p.Logger, broker, conn)
+	} else {
+		p.pluginClient.refresh(broker, conn)
+	}
+
+	return p.pluginClient, nil
+}
+
+func (p *GRPCPluginMedian) ClientConfig() *plugin.ClientConfig {
+	return &plugin.ClientConfig{
+		HandshakeConfig:  PluginMedianHandshakeConfig(),
+		Plugins:          map[string]plugin.Plugin{PluginMedianName: p},
+		AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+	}
 }
 
 var _ PluginMedian = (*pluginMedianClient)(nil)
 
 type pluginMedianClient struct {
-	*lggrBroker
+	*pluginClient
+	*serviceClient
 
-	grpc pb.PluginMedianClient
+	median pb.PluginMedianClient
 }
 
-func newMedianPluginClient(lggr logger.Logger, broker *plugin.GRPCBroker, conn *grpc.ClientConn) *pluginMedianClient {
-	lggr = logger.Named(lggr, "MedianReportingClient")
-	return &pluginMedianClient{lggrBroker: &lggrBroker{lggr, broker}, grpc: pb.NewPluginMedianClient(conn)}
+func newPluginMedianClient(stopCh <-chan struct{}, lggr logger.Logger, broker *plugin.GRPCBroker, conn *grpc.ClientConn) *pluginMedianClient {
+	lggr = logger.Named(lggr, "PluginMedianClient")
+	pc := newPluginClient(stopCh, lggr, broker, conn)
+	return &pluginMedianClient{pluginClient: pc, median: pb.NewPluginMedianClient(pc), serviceClient: newServiceClient(pc.lggrBroker, pc)}
```

Comment:
`&pluginMedianClient{pluginClient: pc, median: pb.NewPluginMedianClient(pc), serviceClient: newServiceClient(pc.lggrBroker, pc)`

this reads like a circle. `pc` is a pluginClient and is the only thing passed to the other fields.

what does it mean to have a pluginMedianClient comprised of a pluginClient and a NewPluginMedianClient? it looks like a circle

and what does it mean to have a serviceClient that is constructed from a *pluginClient? it reads like the service client isn't needed at all.

Reply:
Can you elaborate on this? I'm not following what is circular. We have a custom type which embeds two other types to inherit their methods, and also implements an interface using the `pb` client.